✨ Foundation for Synthesis Methods

SmaranTum opened the pull request:
This PR builds the foundation for implementing various synthesis techniques. To this end, it

- introduces a `TruthTable` class for managing (ir-)reversible truth tables
- a parser for PLA files to specify truth tables in a convenient fashion
- methods to transform reversible truth tables to decision diagrams (forming the basis for DD-based synthesis)
- techniques for encoding irreversible functions into reversible functions (based on a Huffman encoding)

burgholzer reviewed (changes requested):
Thanks for properly setting up the PR now!
Now I am finally able to review what has actually changed.

Unfortunately, there are lots of major concerns that need to be addressed. First and foremost, structure and organization:
- _never_ include `main` functions in library source code files. This is what we have unit tests for and anything related to testing library code should reside there (in a separate file like `test_dd_functionality.cpp`). As can be seen from the CI, including such a function even breaks the CI (coverage decreased to <10% because tests can no longer be detected because of the `main` function).
- it's _never_ a good idea to put everything in a single `.cpp` file. There should always be a header file containing all declarations and a source file containing the definitions.
- opening up the `tt_qmdd.cpp` file in CLion reveals `>30` warnings to be fixed. consider using an IDE for that or run `clang-tidy` yourself.
- prefer to not use `using namespace std` and make things explicit
 
However, the biggest concern is that there is absolutely no need to re-implement QMDDs here. The SyReC library already contains our fully-fledged library for decision diagrams as a dependency (in `extern/qfr/extern/dd_package`). It does not make sense to repeat years of development and reinvent the wheel here.
Small remark: prefer to just use the term *decision diagram* or *DD* instead of *QMDD* as it more closely follows the terminology established in the past few years.

What this PR needs is simple:

- [ ] a suitable data structure for representing truth tables (including IO). In a separate header and source file in the `core` part of syrec.
- [ ] a function with a signature similar to `qc::MatrixDD buildDD(const syrec::TruthTable& tt, std::unique_ptr<dd::Package<>>& dd)` that takes a truth table and an instance of one of our decision diagram packages and produces the corresponding DD. Probably this should also reside in the `core` part of syrec, since it is only later used for synthesis.

SmaranTum commented:
Oh shoot!! Okay, will get on with it.

SmaranTum commented:
> Oh shoot!! Okay, will get on with it.

As suggested, I have created a truth table data structure. The pla parser is yet to be implemented. In addition to the above, a function called buildDD(const syrec::TruthTable& tt, std::unique_ptr<dd::Package<>>& dd) is created. It basically stores the corresponding decision diagram in the dd instance.

burgholzer commented:
> > Oh shoot!! Okay, will get on with it.
> 
> As suggested, I have created a truth table data structure. The pla parser is yet to be implemented. In addition to the above, a function called buildDD(const syrec::TruthTable& tt, std::unique_ptr<dd::Package<>>& dd) is created. It basically stores the corresponding decision diagram in the dd instance.

Thank you for that! Please try to always write tests for your newly implemented functions. As you can see from the CI, `0.0%` of the changes are covered by tests. This also makes it quite hard to go through the changes in the GitHub "Files Changed" view as it shows some lines of code interleaved with the message that these lines are not covered by tests.
I'll go through the changes themselves once tests are in place.

The `.pla` parser can, then, be the next Pull Request once this one is merged.

SmaranTum commented:
As suggested, I have added the unit tests and the coverage is above target. One note, for now, the class truth_table is basically a vector. However, when we start considering don't cares, all the commented section can be removed.

burgholzer commented:
Thanks again!
I had a quick look at the changes, mainly focusing on the truth table structure itself. A couple of general remarks:
 - do not use a leading `_` to indicate members of a class. just give them a proper name
 - try to use [camelCase](https://en.wikipedia.org/wiki/Camel_case) as opposed to [snake_case](https://en.wikipedia.org/wiki/Snake_case) for naming variables
 - unique table hits and lookups are not really a good way to test the package as they might (theoretically) change due to the hash function depending on pointer addresses. Instead proper checks would be to construct the truth table for the identity and then compare it to the identity dd (that can be generated with `dd->makeIdent(n)`), or, in case of the CNOT, with the following DD
 ```c++
#include "dd/FunctionalityConstruction.hpp"
using namespace dd::literals; 
auto qc = QuantumComputation(2);
qc.x(0, 1_pc) // creates CNOT with target q0 and control q1
auto matrix = dd::buildFunctionality(&qc, dd) // builds the underlying matrix decision diagram
 ```

Apart from that, my biggest concern is that your data structure for truth tables is extremely inefficient. Just as an example: any truth table function on up to 6 variables can be represented by a single 64bit integer. In comparison, Your data structure is orders of magnitude larger and, as a consequence, hardly scalable. 
I looked a bit around on the internet and there is a rather neat and optimised TT library in the EPFL logic synthesis toolkit: https://github.com/msoeken/kitty. 
Integrating this library as a submodule of our SyReC project seems like the way to go.
The only thing I am not 100% sure about is how/if the library supports don't cares. But that could be something we build on top of the library.

SmaranTum commented:
Hi Lukas, I have implemented the truthTable data structure similar to the one in RevKit. The project is 88% covered and the rest will be covered once we introduce the don't care terminal nodes. Now, I will be working on the BFS traversal and the parser.

burgholzer commented on src/core/truthTable/truth_table.cpp:
```diff
@@ -0,0 +1,252 @@
+        }
+    }
+
+    truthTable::cube_type append_zero(truthTable::cube_type enc) {
+        enc.push_back(false);
+
+        return enc;
+    }
```
this creates a copy of a potentially large vector once. then adds a single element and then makes another copy when returning.

This should be a member function of the `Cube` class that simply appends a zero and does not return anything.

burgholzer commented on src/core/truthTable/truth_table.cpp:
```diff
@@ -0,0 +1,252 @@
+        return enc;
+    }
+
+    truthTable::cube_type insert_zero(truthTable::cube_type& inNew) {
+        inNew.insert(inNew.begin(), false);
+
+        return inNew;
+    }
+
+    truthTable::cube_type append_one(truthTable::cube_type enc) {
+        enc.push_back(true);
+
+        return enc;
+    }
```
same as above.

burgholzer commented on src/core/truthTable/truth_table.cpp:
```diff
@@ -0,0 +1,252 @@
+                        i.second.push_back(emptyVal);
+                    }
+                }
+            }
+
+            //replace the outputs with the corrsponding encoded outputs
+
+            for (auto& i: inOutCube) {
+                inOutCube[i.first] = encInOut[i.second];
+            }
+        }
+
+        //Append zeros to the input if the output length (either encoded or not) is higher compared to the input
+
+        for (auto& i: inOutCube) {
+            if (i.second.size() > i.first.size()) {
+                for (unsigned j = 0; j <= (i.second.size() > i.first.size()); j++) {
+                    truthTable::cube_type dummy = i.first;
+
+                    auto node = inOutCube.extract(i.first);
+
+                    node.key() = insert_zero(dummy);
+
+                    inOutCube.insert(std::move(node));
+                }
+            }
+        }
+
+        tt.clear();
+
+        //update the truth table
+
+        for (auto& i: inOutCube) {
+            truthTable::cube_type firstVal = i.first;
+
+            truthTable::cube_type secondVal = i.second;
+
+            tt.add_entry(firstVal, secondVal);
+        }
+    }
```
this whole function is a mess with linter warnings sprinkled all over it. no chance to review it like that.

burgholzer commented on src/core/truthTable/truth_table.cpp:
```diff
@@ -0,0 +1,252 @@
+        hufCodes(root->right, append_one(enc), encInOut);
+    }
+
+    void HuffmanCodes(truthTable& tt) {
```
this should be a member function of the truth table class

burgholzer commented on src/core/truthTable/truth_table.cpp:
```diff
@@ -0,0 +1,252 @@
+        return enc;
+    }
+
+    void hufCodes(struct syrec::MinHeapNode* root, truthTable::cube_type enc, truthTable::cube_vector& encInOut) {
```
this should probably be a member function of the `Cube` class

SmaranTum replied:
I dont think I can make this a member function of Cube, This is a recursive function which requires the copy of enc. As a result, append zero and append one can also not be member functions of Cube class

SmaranTum replied:
The only thing I can make member function is insert_zero function

burgholzer commented on src/core/truthTable/truth_table.cpp:
```diff
@@ -0,0 +1,252 @@
+        return enc;
+    }
+
+    void hufCodes(struct syrec::MinHeapNode* root, truthTable::cube_type enc, truthTable::cube_vector& encInOut) {
+        if (!root)
+            return;
+
+        if (root->data.size() != 0) {
+            encInOut.insert({root->data, enc});
+        }
+
+        hufCodes(root->left, append_zero(enc), encInOut);
+
+        hufCodes(root->right, append_one(enc), encInOut);
+    }
```
this function is also sprinkled with linter warnings.

burgholzer reviewed (changes requested):
Thank you Smaran for the update. I just managed to make it through at least some of the code.
You can find very detailed comments for the truth table class below.

At the moment, I did not even focus that much on if it is reasonable what the code is doing, but merely pointing out all sorts of linter warnings and bad practices. My advice to you: please, please, please use some sort of linter to assess the quality of your code when developing.
In addition to the `clang-tidy` checks that are directly integrated in CLion I can recommend https://www.sonarlint.org/ to you, which is available as a plugin for Clion and in addition to warnings, also provides proper reasoning on why something should be changed.
`>90%` of the PR comments below would probably not be there, if it had been used.

SmaranTum commented on src/core/truthTable/truth_table.cpp:
```diff
@@ -0,0 +1,252 @@
+
+        truthTable::cube_vector ioVec = tt.io_cube();
+
+        for (auto it = ioVec.begin();; ++it) {
```
Can you suggest? How I could go about this? Basically we need to go beyond the last element and the simplest implementation is to use iterators right? Or am I missing something here?

SmaranTum commented on src/core/truthTable/truth_table.cpp:
```diff
@@ -0,0 +1,252 @@
+            ++first;
+        }
+
+        for (unsigned i = 0; i < (1u << dcPositions.size()); ++i) {
```
Can you suggest how I can refactor this?

burgholzer commented:
Thanks for your efforts in removing all the linter warnings. This makes it so much easier to work with the code.
I just found some time today to work on this myself a little bit. As a result, I refactored large parts of the code in order to make it cleaner and easier to work with. 

A (probably incomplete) list of notable changes:
 - the `Cubes` class is now much more potent and useful, including all the member functions detailed in the comments above
 - DD construction code is moved to the `TruthTable` class since it is better suited as a member function there as opposed to a free function in a separate file.
 - refactored the tests
 - the Hoffman encoding got a major rework to be more C++ like and avoid manually handling memory and memory leaks
 - the truth table extension code was reworked

The last two features are not yet tested. The previously existing tests just called the functions and did not test their actual behavior, which is why I removed the respective calls.
Consider writing separate tests (in a separate file) for the truth table extension and encoding.

SmaranTum commented:
Made minor changes to the TruthTable class i.e., modify contents of cube class, added additional operations and member functions required for testing purposes.

Tested the extent(), encodeHuffman() and augmentWithConstants() functions.

Added dd_synthesis file, which includes functions that are required for the synthesis of any given dd (for now synthesis of reversible functions (without Don't cares) is allowed).

Tested functions of dd_synthesis as well.

burgholzer commented on src/core/truthTable/truth_table.cpp:
```diff
@@ -11,19 +11,21 @@ namespace syrec {
-            auto zero         = Cube(cube);
-            zero[pos]         = false;
+            Cube zero(cube);
+            zero.at(pos) = false;
```
Is there a reason for these lines being changed?

SmaranTum replied:
zero[pos] was never modifying the element in the specific position since [] operator returns a const. As a result, it ended up being a infinite recursion. Since vectors have an at method, I thought I could define an at method as well

burgholzer commented on src/core/truthTable/truth_table.cpp:
```diff
@@ -11,19 +11,21 @@ namespace syrec {
+
+            Cube one(cube);
+            one.at(pos) = true;
+
```
same question here

burgholzer commented:
Thanks Smaran,

I'll try to go through the changes as fast as possible and get back to you.

One thing I noticed immediately is that you seem to have committed a different version of the QFR submodule.
There should be no reason for that change.

SmaranTum commented:
> Thanks Smaran,
> 
> I'll try to go through the changes as fast as possible and get back to you.
> 
> One thing I noticed immediately is that you seem to have committed a different version of the QFR submodule. There should be no reason for that change.

I actually don't know how I ended up changing the version (I never intended to change). How do I change it back?

burgholzer commented:
> > Thanks Smaran,
> > I'll try to go through the changes as fast as possible and get back to you.
> > One thing I noticed immediately is that you seem to have committed a different version of the QFR submodule. There should be no reason for that change.
> 
> I actually don't know how I ended up changing the version (I never intended to change). How do I change it back?

I just merged the `main` branch into the PR. Conflicts are now resolved and CI is running again. (Still haven't had time to look through the changes though)

burgholzer commented on include/core/truthTable/pla_parser.hpp:
```diff
@@ -0,0 +1,39 @@
+
+namespace syrec {
+
+    [[maybe_unused]] inline syrec::TruthTable::Cube::Value transform_pla_to_constants(const char& c) {
+        syrec::TruthTable::Cube::Value constant;
+        syrec::TruthTable::Cube::Value one  = true;
+        syrec::TruthTable::Cube::Value zero = false;
+        switch (c) {
+            case '-':
+            case '~':
+                return constant;
+
+            case '0':
+                return zero;
+            case '1':
+                return one;
+
+            default:
+                assert(false);
+                return constant;
+        }
+    }
```
Remove the `[[maybe unused]]`.

This function can be simplified. See the below suggestion.
```suggestion
    inline TruthTable::Cube::Value transform_pla_to_constants(const char& c) {
        switch (c) {
            case '-':
            case '~':
                return {};
            case '0':
                return false;
            case '1':
                return true;
            default:
                assert(false);
                return {};
        }
    }
```

Note that the `assert(false)` only triggers in debug mode and not in release configurations. If you want to indicate an error, then throw an exception.

burgholzer commented on include/core/truthTable/pla_parser.hpp:
```diff
@@ -0,0 +1,39 @@
+        }
+    }
+
+    [[maybe_unused]] inline std::string trim(std::string& str) {
```
remove the `[[maybe_unused]]`.

burgholzer commented on include/core/truthTable/pla_parser.hpp:
```diff
@@ -0,0 +1,39 @@
+        return str;
+    }
+
+    [[maybe_unused]] void pla_parser(std::istream& in, syrec::TruthTable& reader);
```
remove the `[[maybe_unused]]`.

burgholzer replied:
use verbs for function names instead of nouns. this function is not a parser, it parses a file, i.e., call it `parse_pla`.

burgholzer replied:
why is the truth table called `reader` here and `spec` everywhere else? try to stick to consistent naming.

burgholzer replied:
Drop the `syrec::` qualifier. You are already in the syrec namespace. Please check throughout this file and the corresponding source file.

burgholzer commented on include/core/truthTable/pla_parser.hpp:
```diff
@@ -0,0 +1,39 @@
+        return str;
+    }
+
+    [[maybe_unused]] void pla_parser(std::istream& in, syrec::TruthTable& reader);
+
+    [[maybe_unused]] void read_pla(TruthTable& spec, std::istream& in);
```
remove the `[[maybe_unused]]`.

burgholzer replied:
Please use consistent ordering of function parameters, e.g., `truthTable` should always come first.

burgholzer commented on include/core/truthTable/pla_parser.hpp:
```diff
@@ -0,0 +1,39 @@
+
+namespace syrec {
+
+    [[maybe_unused]] inline syrec::TruthTable::Cube::Value transform_pla_to_constants(const char& c) {
```
this function does not seem to be appropriately named since it does not only return constants but also `true` and `false`.

burgholzer commented on src/core/truthTable/pla_parser.cpp:
```diff
@@ -0,0 +1,91 @@
+
+        pla_parser(in, spec);
+
+        spec.extend();
+
+        spec.encodeHuffman();
+
+        spec.augmentWithConstants();
```
these calls should not be part of this function. there is no reason why reading a PLA file should also perform encoding and modify the truth table.
This function should merely read in the truth table from a PLA file and nothing more.
As such, this function is superfluous and can be removed. Just inline the remaining two calls in the other `read_pla` function.

burgholzer commented on src/core/truthTable/pla_parser.cpp:
```diff
@@ -0,0 +1,91 @@
+            line = std::regex_replace(line, whitespace, std::string(" "));
+            if ((line.empty()) || (line.rfind('#', 0) == 0) || (line.rfind(".i", 0) == 0) || (line.rfind(".o", 0) == 0) || (line.rfind(".ilb", 0) == 0) || (line.rfind(".ob", 0) == 0) || (line.rfind(".p", 0) == 0) || (line.rfind(".type ", 0) == 0)) { continue; }
+
+            //std::string comment( std::find_if( line.begin(), line.end(), []( char c ) { return c != '#'; } ) );
+            //reader.on_comment( comment );
+
+            //reader.on_num_inputs( static_cast<unsigned>( line.substr( 3 ) ) );
+
+            //reader.on_num_outputs( boost::lexical_cast<unsigned>( line.substr( 3 ) ) );
+
+            //reader.on_num_products( boost::lexical_cast<unsigned>( line.substr( 3 ) ) );
+
+            //reader.on_type( line.substr( 6 ) );
```
please remove any unused/commented lines from this file.

burgholzer commented on src/core/truthTable/pla_parser.cpp:
```diff
@@ -0,0 +1,91 @@
+            //reader.on_type( line.substr( 6 ) );
+            else if (line == ".e") {
+                break;
+                //reader.on_end();
```
please remove any unused/commented lines from this file.

burgholzer commented on src/core/truthTable/pla_parser.cpp:
```diff
@@ -0,0 +1,91 @@
+                assert(!line.empty() && (line[0] == '0' || line[0] == '1' || line[0] == '-'));
+
+                std::vector<std::string> inout;
+                //boost::split( inout, line, boost::is_any_of(" \t|" ), boost::token_compress_on );
```
please remove any unused/commented lines from this file.

burgholzer commented on src/core/truthTable/pla_parser.cpp:
```diff
@@ -0,0 +1,91 @@
+        std::regex  whitespace("\\s+");
+
+        while (in.good() && getline(in, line)) {
+            trim(line);
```
you defined the `trim` function to return the trimmed string and do not use it here. Consider making the `trim` function return `void`.

burgholzer commented on src/core/truthTable/pla_parser.cpp:
```diff
@@ -0,0 +1,91 @@
+
+        while (in.good() && getline(in, line)) {
+            trim(line);
+            line = std::regex_replace(line, whitespace, std::string(" "));
```
```suggestion
            line = std::regex_replace(line, whitespace, " ");
```

burgholzer commented on src/core/truthTable/pla_parser.cpp:
```diff
@@ -0,0 +1,91 @@
+        while (in.good() && getline(in, line)) {
+            trim(line);
+            line = std::regex_replace(line, whitespace, std::string(" "));
+            if ((line.empty()) || (line.rfind('#', 0) == 0) || (line.rfind(".i", 0) == 0) || (line.rfind(".o", 0) == 0) || (line.rfind(".ilb", 0) == 0) || (line.rfind(".ob", 0) == 0) || (line.rfind(".p", 0) == 0) || (line.rfind(".type ", 0) == 0)) { continue; }
```
why does the parser ignore all "frontmatter" declarations? couldn't these contain useful information?

SmaranTum replied:
To implement our functionality I don't think its required. Later on if we need any additional information, we can use them.

burgholzer commented on src/core/truthTable/pla_parser.cpp:
```diff
@@ -0,0 +1,91 @@
+            }
+
+            else {
+                assert(!line.empty() && (line[0] == '0' || line[0] == '1' || line[0] == '-'));
```
this does not handle the tilde character that is handled in the other parser function.

burgholzer commented on src/core/truthTable/pla_parser.cpp:
```diff
@@ -0,0 +1,91 @@
+
+                std::vector<std::string> inout;
+                //boost::split( inout, line, boost::is_any_of(" \t|" ), boost::token_compress_on );
+                const int MAX = 200;
+                char      input[MAX];
+                strcpy(input, line.c_str());
+                char* token = std::strtok(input, " \t|");
+                while (token != nullptr) {
+                    inout.emplace_back(token);
+                    token = std::strtok(nullptr, " ");
+                }
```
this is unnecessarily complex and not very C++ like. this would be much better with something like
```c++
auto ss = std::stringstream(line);
std::string s;
while (std::getline(ss, s, " ")) {
  inout.emplace_back(s);
}
```

burgholzer commented on src/core/truthTable/pla_parser.cpp:
```diff
@@ -0,0 +1,91 @@
+                    token = std::strtok(nullptr, " ");
+                }
+
+                assert(inout.size() == 2);
```
Again, asserts are only triggered in debug builds. If this is intended, then it's fine. If not, then throw an appropriate error here.

burgholzer commented on test/unittests/test_pla_parser.cpp:
```diff
@@ -0,0 +1,80 @@
+TEST_F(PlaParserTest, andTest) {
+    std::string circAnd = test_circuits_dir + "and.pla";
+
+    EXPECT_EQ(testPla.nInputs(), 0U);
+    EXPECT_EQ(testPla.nOutputs(), 0U);
+    EXPECT_EQ(testPla.ioCube().size(), 0U);
```
these checks are superfluous and can be removed in all of the tests as they will always be satisfied and test nothing.

burgholzer commented on test/unittests/test_pla_parser.cpp:
```diff
@@ -0,0 +1,80 @@
+
+    EXPECT_TRUE(read_pla(testPla, circAnd));
+
+    EXPECT_EQ(testPla.nInputs(), 3U);
+    EXPECT_EQ(testPla.nOutputs(), 3U);
+    EXPECT_EQ(testPla.ioCube().size(), 4U);
```
These checks will have to be adapted based on the changes requested above. Merely reading a PLA file should not involve any embedding or encoding.

If you want to test that behaviour as well, then write separate and explicit tests for that.

burgholzer replied:
These tests should also verify that the PLA file has been parsed correctly. Just checking whether the number of inputs and outputs are correct is not sufficient.

burgholzer commented on test/unittests/test_pla_parser.cpp:
```diff
@@ -0,0 +1,80 @@
+
+using namespace syrec;
+
+class PlaParserTest: public testing::TestWithParam<std::string> {
```
why is this a parameterised test suite if there are no parameterised tests?

burgholzer commented on test/CMakeLists.txt:
```diff
@@ -15,6 +15,11 @@ if (NOT TARGET gtest OR NOT TARGET gmock)
 	endif ()
 endif ()
 
+#add_executable(${PROJECT_NAME}_example ${CMAKE_CURRENT_SOURCE_DIR}/test.cpp)
+#target_link_libraries(${PROJECT_NAME}_example PRIVATE ${PROJECT_NAME})
+#set_target_properties(${PROJECT_NAME}_example PROPERTIES FOLDER tests CMAKE_CXX_STANDARD_REQUIRED ON CXX_EXTENSIONS OFF)
+#enable_lto(${PROJECT_NAME}_example)
+
```
why is this commented out here?

burgholzer reviewed (changes requested):
This review only covers the PLA parser part. I'll cover the DD synthesis part in a separate review.

For the future: Please try to limit pull requests to single additions of features. This PR now adds at least four different features. Not only does this make it hard to review, it also drags out the review process and, hence, the integration of new features.

Please also update the PR title and its description to properly reflect what this PR adds.

burgholzer commented on include/core/truthTable/truth_table.hpp:
```diff
@@ -0,0 +1,211 @@
+
+        auto augmentWithConstants() -> void;
+
+        auto buildDD(std::unique_ptr<dd::Package<>>& dd) const -> dd::mEdge;
```
This function should not be a member of the `TruthTable` class. It should reside somewhere in a file in the `include/algorithms/synthesis` directory. Possibly something like `truth_table_to_dd.hpp`. 
Or, probably even better, it should be in the `dd_synthesis` file.

Note that this requires changing the function signature to
```c++
auto buildDD(const TruthTable& tt, std::unique_ptr<dd::Package<>>& dd) const -> dd::mEdge;
```

burgholzer commented on include/core/truthTable/truth_table.hpp:
```diff
@@ -0,0 +1,211 @@
+
+        auto extend() -> void;
+
+        auto encodeHuffman() -> void;
```
In the same spirit as the comment on the `buildDD` function, this function should also not be placed here.
"Encoding" is certainly something that fits in the `algorithms` directory. 
Probably in `include/algorithms/synthesis/encoding.hpp`.

This is also where the `MinHeapNode` class should be moved to as it is not inherently related to the TruthTable data structure.

burgholzer commented on include/core/truthTable/truth_table.hpp:
```diff
@@ -0,0 +1,211 @@
+            return cubeMap;
+        }
+
+        auto extend() -> void;
```
It's debatable whether this function and the `augmentWithConstants` function should be here. I'd rather consider them "algorithms" than "core functionality".

burgholzer commented on test/unittests/test_dd_synthesis.cpp:
```diff
@@ -0,0 +1,551 @@
+    TruthTable::Cube c111111{};
+
+    void SetUp() override {
+        dd2Bit = std::make_unique<dd::Package<>>(2U);
+        dd3Bit = std::make_unique<dd::Package<>>(3U);
+        dd4Bit = std::make_unique<dd::Package<>>(4U);
+        dd6Bit = std::make_unique<dd::Package<>>(6U);
```
these can be moved to the definition of the `dd` packages. Furthermore, it is sufficient to just allocate a single DD package that can hold the maximum number of lines you want to test with, i.e., 
```c++
dd = std::make_unique<dd::Package<>>(6U);
```

burgholzer commented on test/unittests/test_dd_synthesis.cpp:
```diff
@@ -0,0 +1,551 @@
+    TruthTable::Cube c011010{};
+    TruthTable::Cube c011011{};
+    TruthTable::Cube c011100{};
+    TruthTable::Cube c011101{};
+    TruthTable::Cube c011110{};
+    TruthTable::Cube c011111{};
+
+    TruthTable::Cube c100000{};
+    TruthTable::Cube c100001{};
+    TruthTable::Cube c100010{};
+    TruthTable::Cube c100011{};
+    TruthTable::Cube c100100{};
+    TruthTable::Cube c100101{};
+    TruthTable::Cube c100110{};
+    TruthTable::Cube c100111{};
+    TruthTable::Cube c101000{};
+    TruthTable::Cube c101001{};
+    TruthTable::Cube c101010{};
+    TruthTable::Cube c101011{};
+    TruthTable::Cube c101100{};
+    TruthTable::Cube c101101{};
+    TruthTable::Cube c101110{};
+    TruthTable::Cube c101111{};
+
+    TruthTable::Cube c110000{};
+    TruthTable::Cube c110001{};
+    TruthTable::Cube c110010{};
+    TruthTable::Cube c110011{};
+    TruthTable::Cube c110100{};
+    TruthTable::Cube c110101{};
+    TruthTable::Cube c110110{};
+    TruthTable::Cube c110111{};
+    TruthTable::Cube c111000{};
+    TruthTable::Cube c111001{};
+    TruthTable::Cube c111010{};
+    TruthTable::Cube c111011{};
+    TruthTable::Cube c111100{};
+    TruthTable::Cube c111101{};
+    TruthTable::Cube c111110{};
+    TruthTable::Cube c111111{};
```
This is truly horrible. Now that you implemented a PLA parser, it is far easier to specify the truth tables in PLA format as strings in the respective tests. Something along the lines of (untested):
```
auto pla = "... pla definition";
auto ss = std::stringstream(pla);
TruthTable tt{};
read_pla(tt, ss);
```

burgholzer replied:
I understand that. But since the PLA parser is implemented now, this can be done properly.

burgholzer commented on test/unittests/test_extend_tt.cpp:
```diff
@@ -0,0 +1,140 @@
+    void SetUp() override {
+        dd = std::make_unique<dd::Package<>>(nqubits);
+
+        c00 = TruthTable::Cube::fromInteger(0b00U, 2U);
+        c01 = TruthTable::Cube::fromInteger(0b01U, 2U);
+        c10 = TruthTable::Cube::fromInteger(0b10U, 2U);
+        c11 = TruthTable::Cube::fromInteger(0b11U, 2U);
+
+        c000 = TruthTable::Cube::fromInteger(0b000U, 3U);
+        c001 = TruthTable::Cube::fromInteger(0b001U, 3U);
+        c010 = TruthTable::Cube::fromInteger(0b010U, 3U);
+        c011 = TruthTable::Cube::fromInteger(0b011U, 3U);
+        c100 = TruthTable::Cube::fromInteger(0b100U, 3U);
+        c101 = TruthTable::Cube::fromInteger(0b101U, 3U);
+        c110 = TruthTable::Cube::fromInteger(0b110U, 3U);
+        c111 = TruthTable::Cube::fromInteger(0b111U, 3U);
```
now that you implemented the PLA parser, this should be refactored appropriately. See the comment on the DD synthesis test file for more information.

SmaranTum replied:
have removed the most of the cubes, however, I have kept a few for checking whether particular cube instances are added or not.

burgholzer commented on test/unittests/test_huffman.cpp:
```diff
@@ -0,0 +1,251 @@
+
+    TruthTable::Cube::Value emptyVal;
+
+    TruthTable::Cube c00{};
+    TruthTable::Cube c01{};
+    TruthTable::Cube c10{};
+    TruthTable::Cube c11{};
+
+    TruthTable::Cube c000{};
+    TruthTable::Cube c001{};
+    TruthTable::Cube c010{};
+    TruthTable::Cube c011{};
+    TruthTable::Cube c100{};
+    TruthTable::Cube c101{};
+    TruthTable::Cube c110{};
+    TruthTable::Cube c111{};
+
+    TruthTable::Cube c_11{};
+    TruthTable::Cube c1_0{};
+
+    TruthTable::Cube c0_{};
+    TruthTable::Cube c1_{};
+
+    TruthTable::Cube c0__{};
+    TruthTable::Cube c1__{};
+
+    TruthTable::Cube c10_{};
```
similar to the other two test files, this should be refactored to make use of the PLA parser.

SmaranTum replied:
I have removed the most of the cubes, however, I have kept a few for checking whether particular cube instances are added or not (outputs of a Pla aren't supposed to have don't cares?).

burgholzer commented on test/unittests/test_dd_synthesis.cpp:
```diff
@@ -0,0 +1,551 @@
+using namespace dd::literals;
+using namespace syrec;
+
+class DDSynth: public testing::TestWithParam<std::string> {
```
Again, why is this a parameterised test suite without parameterised tests?
Also: The class name should probably contain the word `test`.

burgholzer commented on test/unittests/test_dd_synthesis.cpp:
```diff
@@ -0,0 +1,551 @@
+
+    EXPECT_NE(ttDD, dd3Bit->makeIdent(3U));
+
+    auto src = ttDD;
+
+    EXPECT_NE(src, dd3Bit->makeIdent(3U));
```
why copy the DD here and repeat the same check.
remove the copy and the redundant check throughout this file.

burgholzer commented on test/unittests/test_dd_synthesis.cpp:
```diff
@@ -0,0 +1,551 @@
+
+    EXPECT_NE(src, dd3Bit->makeIdent(3U));
+
+    algoQ(src, dd3Bit);
+
+    EXPECT_EQ(src, dd3Bit->makeIdent(3U));
```
while it is good to know that the synthesis is able to transform the DD to the identity, this alone is not sufficient for ensuring that the synthesis works as intended. For this, the resulting circuit has to be checked. Adjust these tests accordingly.

burgholzer reviewed (changes requested):
These are mostly general remarks on the additions in this PR and how tests are written.

I have not looked at the synthesis code itself because it is far from being ready for review. A couple of reasons:
- The `dd_synthesis` files should not reside in the `core` folder, but rather in the `algorithms/synthesis` folder
- The function names alone are completely incomprehensible and undocumented. What should `algoQ` mean? `algoP`? where do the numbers in the function names come from? how does this relate to synthesis?
- You are proposing synthesis methods, yet your methods do not return the synthesized circuit. As such, they do nothing. A proper signature for a synthesis function would be
```c++
auto synthesize(dd::mEdge& src, std::unique_ptr<dd::Package<>>& dd) -> qc::QuantumComputation
```
- The synthesis functions still contain lots of debug code (print statements) that should be removed.

Since these might be huge changes and will require even more time, I would suggest that you remove any DD synthesis related code from this PR, work on the rest of the requested changes in this PR so that it can be merged, and then create a separate PR for the DD synthesis functionality.

SmaranTum commented:
> These are mostly general remarks on the additions in this PR and how tests are written.
> 
> I have not looked at the synthesis code itself because it is far from being ready for review. A couple of reasons:
> 
>     * The `dd_synthesis` files should not reside in the `core` folder, but rather in the `algorithms/synthesis` folder
> 
>     * The function names alone are completely incomprehensible and undocumented. What should `algoQ` mean? `algoP`? where do the numbers in the function names come from? how does this relate to synthesis?
> 
>     * You are proposing synthesis methods, yet your methods do not return the synthesized circuit. As such, they do nothing. A proper signature for a synthesis function would be
> 
> 
> ```c++
> auto synthesize(dd::mEdge& src, std::unique_ptr<dd::Package<>>& dd) -> qc::QuantumComputation
> ```
> 
>     * The synthesis functions still contain lots of debug code (print statements) that should be removed.
> 
> 
> Since these might be huge changes and will require even more time, I would suggest that you remove any DD synthesis related code from this PR, work on the rest of the requested changes in this PR so that it can be merged, and then create a separate PR for the DD synthesis functionality.

so basically for the dd synthesis part I referred to the 
[12aspdac_qmdd_synth_rev.pdf](https://github.com/cda-tum/syrec/files/9444677/12aspdac_qmdd_synth_rev.pdf) In which AlgoQ and Algo P were used.

Also, I know the synthesis methods do not return anything, however, the respective functions change the truth table transformed dd to Identity. I thought I would ask you the question on how the output of the synthesis should be represented and based on your suggestion I will use the QunatumComputation object itself

burgholzer commented:
> > These are mostly general remarks on the additions in this PR and how tests are written.
> > I have not looked at the synthesis code itself because it is far from being ready for review. A couple of reasons:
> > ```
> > * The `dd_synthesis` files should not reside in the `core` folder, but rather in the `algorithms/synthesis` folder
> > 
> > * The function names alone are completely incomprehensible and undocumented. What should `algoQ` mean? `algoP`? where do the numbers in the function names come from? how does this relate to synthesis?
> > 
> > * You are proposing synthesis methods, yet your methods do not return the synthesized circuit. As such, they do nothing. A proper signature for a synthesis function would be
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```c++
> > auto synthesize(dd::mEdge& src, std::unique_ptr<dd::Package<>>& dd) -> qc::QuantumComputation
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```
> > * The synthesis functions still contain lots of debug code (print statements) that should be removed.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Since these might be huge changes and will require even more time, I would suggest that you remove any DD synthesis related code from this PR, work on the rest of the requested changes in this PR so that it can be merged, and then create a separate PR for the DD synthesis functionality.
> 
> so basically for the dd synthesis part I referred to the [12aspdac_qmdd_synth_rev.pdf](https://github.com/cda-tum/syrec/files/9444677/12aspdac_qmdd_synth_rev.pdf) In which AlgoQ and Algo P were used.
> 
> Also, I know the synthesis methods do not return anything, however, the respective functions change the truth table transformed dd to Identity. I thought I would ask you the question on how the output of the synthesis should be represented and based on your suggestion I will use the QunatumComputation object itself

While that is true, the paper also gives proper names to these algorithms (QMDD-based Synthesis in case of Algorithm Q).
Algorithm P and all its steps also have proper names in the paper.

The output of the synthesis should be a valid `QuantumComputation` object 👍🏼

burgholzer commented on include/core/io/pla_parser.hpp:
```diff
@@ -0,0 +1,31 @@
+        str.erase(0, str.find_first_not_of(' ')); //prefixing spaces
+    }
+
+    void parse_pla(TruthTable& reader, std::istream& in);
```
`reader` seems off to me as a name. The truth table does not "read" anything. Why not simply call it `tt`?
Same goes for the function below.

burgholzer commented on src/core/io/pla_parser.cpp:
```diff
@@ -0,0 +1,68 @@
+
+namespace syrec {
+
+    [[maybe_unused]] void parse_pla(TruthTable& reader, std::istream& in) {
```
remove the `[[maybe_unused]]`

burgholzer commented on src/core/io/pla_parser.cpp:
```diff
@@ -0,0 +1,68 @@
+            }
+
+            else {
+                assert(!line.empty() && (line[0] == '0' || line[0] == '1' || line[0] == '-' || line[0] == '~'));
```
the `!line.empty()` condition can be removed as it can never be false.

burgholzer commented on src/core/io/pla_parser.cpp:
```diff
@@ -0,0 +1,68 @@
+                    inout.emplace_back(tokenString);
+                }
+
+                if (inout.size() != 2)
+                    std::cerr << "I/O not available" << std::endl;
```
This is much worse than the `assert` that was here before. Now the execution continues in the case of an error even in the debug configuration.
Either keep the assert from before or throw an exception here.

burgholzer commented on test/CMakeLists.txt:
```diff
@@ -15,11 +15,6 @@ if (NOT TARGET gtest OR NOT TARGET gmock)
-#target_link_libraries(${PROJECT_NAME}_example PRIVATE ${PROJECT_NAME})
-#set_target_properties(${PROJECT_NAME}_example PROPERTIES FOLDER tests CMAKE_CXX_STANDARD_REQUIRED ON CXX_EXTENSIONS OFF)
-#enable_lto(${PROJECT_NAME}_example)
-
```
why remove these lines? the test example was just fine.
If you really want to get rid of the CMake target, then also remove the corresponding file from the repository.

SmaranTum replied:
I don't think I have added any test.cpp file.

burgholzer replied:
Interesting. Then I really don't know where this was coming from in the beginning. Consider this resolved then.

burgholzer commented on test/unittests/test_pla_parser.cpp:
```diff
@@ -1,80 +1,118 @@
+
+    EXPECT_TRUE(it10 != testPla.ioCube().end());
+    EXPECT_EQ(it10->second, c1);
 }
```
do these examples add anything more to the test suite? or do the other tests already suffice?
if not, then consider removing them and the respective files.

burgholzer reviewed (changes requested):
Just some quick feedback at the end of the day.

burgholzer commented on include/algorithms/synthesis/encoding.hpp:
```diff
@@ -0,0 +1,159 @@
+
+    inline auto extend(TruthTable& tt) -> void {
+        // ensure that the resulting complete table can be stored in the cube map (at most 63 inputs, probably less in practice)
+        assert(tt.cubeMap.empty() || (tt.nInputs() <= std::log2(tt.cubeMap.max_size()) && tt.nInputs() <= 63U));
```
this assert only triggers for debug configurations. however, you want to make sure that an overflow can never happen.
convert this to proper checks outside of an `assert` and throw an exception if the conditions are not satisfied.

SmaranTum replied:
done 👍🏼

burgholzer commented on include/algorithms/synthesis/encoding.hpp:
```diff
@@ -0,0 +1,159 @@
+#ifndef ENCODING_HPP
+#define ENCODING_HPP
```
try to remain consistent with header guards/`#pragma once` when introducing new files.
I'd personally prefer `#pragma once` everywhere over header guards.

SmaranTum replied:
done 👍🏼

burgholzer commented on src/algorithms/synthesis/dd_synthesis.cpp:
```diff
@@ -0,0 +1,57 @@
+        return dd->makeDDNode(label, edges);
+    }
+} // namespace syrec
+  // namespace syrec
```
this line is redundant
```suggestion
```

SmaranTum replied:
done 👍🏼

burgholzer commented on src/algorithms/synthesis/dd_synthesis.cpp:
```diff
@@ -0,0 +1,57 @@
+#include "algorithms/synthesis/dd_synthesis.hpp"
+
+#include "dd/FunctionalityConstruction.hpp"
+using namespace dd::literals;
```
I don't quite see where this is necessary. Consider removing it.

SmaranTum replied:
done 👍🏼

burgholzer commented on test/unittests/test_extend_tt.cpp:
```diff
@@ -0,0 +1,121 @@
+    TruthTable::Cube c111{};
+
+    void SetUp() override {
+        c00 = TruthTable::Cube::fromInteger(0b00U, 2U);
+        c11 = TruthTable::Cube::fromInteger(0b11U, 2U);
+
+        c000 = TruthTable::Cube::fromInteger(0b000U, 3U);
+        c001 = TruthTable::Cube::fromInteger(0b001U, 3U);
+        c010 = TruthTable::Cube::fromInteger(0b010U, 3U);
+        c011 = TruthTable::Cube::fromInteger(0b011U, 3U);
+        c100 = TruthTable::Cube::fromInteger(0b100U, 3U);
+        c101 = TruthTable::Cube::fromInteger(0b101U, 3U);
+        c110 = TruthTable::Cube::fromInteger(0b110U, 3U);
+        c111 = TruthTable::Cube::fromInteger(0b111U, 3U);
```
I'd propose to introduce a
```c++
auto equals(std::uint64_t num) -> bool
```
method in the `Cube` class.
Additionally, a
```c++
auto find(std::uint64_t num) -> decltype(cubeMap.begin())
```
method in the `TruthTable` class might be helpful that creates a corresponding cube and calls `find()` on the underlying `cubeMap`.

Then all of these variables can be removed (also in the other test files) and the respective `EXPECT_EQ(cube, cXYZ)` calls could just be replaced with `EXPECT_TRUE(cube.equals(0bXYZU))`.

SmaranTum replied:
done 👍🏼

burgholzer commented on include/core/truthTable/truth_table.hpp:
```diff
@@ -0,0 +1,170 @@
+            cubeMap.clear();
+        }
+
+        [[nodiscard]] const CubeMap& ioCube() const {
+            return cubeMap;
+        }
```
this function should be removed. the `cubeMap` should not be accessible from outside the class except by using the public interface of the `TruthTable` class.

SmaranTum replied:
done 👍🏼

burgholzer commented on test/unittests/test_huffman.cpp:
```diff
@@ -0,0 +1,152 @@
+        c010 = TruthTable::Cube::fromInteger(0b010U, 3U);
+        c011 = TruthTable::Cube::fromInteger(0b011U, 3U);
+
+        c0_.emplace_back(false);
+        c0_.emplace_back(emptyVal);
+
+        c0__.emplace_back(false);
+        c0__.emplace_back(emptyVal);
+        c0__.emplace_back(emptyVal);
+
+        c1__.emplace_back(true);
+        c1__.emplace_back(emptyVal);
+        c1__.emplace_back(emptyVal);
```
Similar to the `Cube::fromInteger` method there should be a `Cube::fromString` method that can handle don't cares. Then constructions like these are considerably simplified.

Furthermore, it might make sense to not only provide the `equals()` and `find()` functions mentioned in the other comment for integers, but also for strings.

This allows to remove all these variable definitions here as well.

SmaranTum replied:
done 👍🏼

burgholzer reviewed (changes requested):
Overall this looks mostly good. The requested changes mainly fall into two categories

- C++ coding guidelines that should be followed
- additional functionality that improves the usability of the developed classes and allows to improve the tests

Two other comments on the overall workflow:

- please do not resolve conversations yourself in general. If a requested change is incorporated in a new commit, then simply reply with something like "done 👍🏼" on the respective comment. That way, anyone related to pull requests of this project (at the moment mostly me) gets a notification. This makes it easier to keep track of progess.
- please re-request a review on GitHub once all change requests have been incorporated. Only then it becomes clear that you are satisfied with your changes and that you'd like to have them reviewed again.

SmaranTum commented on include/algorithms/synthesis/encoding.hpp:
```diff
@@ -0,0 +1,159 @@
+        }
+    };
+
+    inline auto extend(TruthTable& tt) -> void {
```
done 👍🏼

SmaranTum commented on include/core/truthTable/truth_table.hpp:
```diff
@@ -0,0 +1,170 @@
+            auto reserve(const std::size_t n) -> void {
+                cube.reserve(n);
+            }
+            auto resize(const std::size_t n, const Value& val = Value()) -> void {
+                cube.resize(n, val);
+            }
```
done 👍🏼 (I think these methods wee defined by you right?)

burgholzer replied:
I believe so. Self-review can be helpful some times ;o)

Feel free to resolve this conversation.

SmaranTum commented on src/core/io/pla_parser.cpp:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    bool read_pla(TruthTable& tt, const std::string& filename) {
+        std::ifstream is;
+        is.open(filename.c_str(), std::ifstream::in);
```
done 👍🏼

SmaranTum commented on src/core/io/pla_parser.cpp:
```diff
@@ -0,0 +1,68 @@
+                }
+
+                TruthTable::Cube cubeOut;
+                cubeOut.reserve(inout[1].size());
```
done 👍🏼

SmaranTum commented on src/core/io/pla_parser.cpp:
```diff
@@ -0,0 +1,68 @@
+                    throw std::invalid_argument("Invalid I/O");
+
+                TruthTable::Cube cubeIn;
+                cubeIn.reserve(inout[0].size());
```
done 👍🏼

SmaranTum commented on src/core/io/pla_parser.cpp:
```diff
@@ -0,0 +1,68 @@
+        while (in.good() && getline(in, line)) {
+            trim(line);
+            line = std::regex_replace(line, whitespace, " ");
+            if ((line.empty()) || (line.rfind('#', 0) == 0) || (line.rfind(".i", 0) == 0) || (line.rfind(".o", 0) == 0) || (line.rfind(".ilb", 0) == 0) || (line.rfind(".ob", 0) == 0) || (line.rfind(".p", 0) == 0) || (line.rfind(".type ", 0) == 0)) {
+                continue;
+            }
```
done 👍🏼

SmaranTum commented on src/core/io/pla_parser.cpp:
```diff
@@ -0,0 +1,68 @@
+                }
+
+                if (inout.size() != 2)
+                    throw std::invalid_argument("Invalid I/O");
```
done 👍🏼

SmaranTum commented on include/core/truthTable/truth_table.hpp:
```diff
@@ -0,0 +1,170 @@
+                return cube[pos];
+            }
+
+            auto operator[](std::size_t pos) const -> Value {
```
removed at operator since it is not used 

done 👍🏼

burgholzer commented on include/core/truthTable/truth_table.hpp:
```diff
@@ -17,6 +17,8 @@
 
 namespace syrec {
 
+    auto transformCharToCubeValue(const char& c) -> std::optional<bool>;
```
I'd propose to make this function a static member function of the `Cube`class in order to avoid it being a free floating function. Something like
```c++
static auto getValue(const char& c) -> Value {
  ...
}
```

SmaranTum replied:
done 👍🏼

burgholzer commented on include/core/truthTable/truth_table.hpp:
```diff
@@ -50,6 +52,18 @@ namespace syrec {
+
+            // construct a cube from a string
+            static auto fromString(const std::string& str) -> Cube {
+                assert(str.size() <= 64U);
```
why the limitation to less than 64 bits? Strings in particular are the way to specify cubes with more than 64bits.

SmaranTum replied:
done 👍🏼

burgholzer commented on include/core/truthTable/truth_table.hpp:
```diff
@@ -82,12 +96,13 @@ namespace syrec {
-                return cube[pos];
+                const std::vector<Value>& x = cube;
+                return x[pos];
             }
```
This is not a proper definition of the const version of `operator[]`. It should be
```c++
auto operator[](std::size_t pos) -> const Value& {
  return cube[pos];
}
```
Especially, it should not return a value, but a const reference and there is no need to create a temporary.

SmaranTum replied:
I had tried this earlier, however, I was getting the following error:

Functions that differ only in their return type cannot be overloaded

What do you suggest I do?

burgholzer replied:
I just edited the comment above. The suggestion was missing a const.
That way it should work because it is a different signature.

SmaranTum replied:
done 👍🏼

burgholzer commented on include/core/truthTable/truth_table.hpp:
```diff
@@ -112,6 +133,16 @@ namespace syrec {
                 cube.pop_back();
             }
 
+            [[nodiscard]] auto equals(const std::uint64_t num) const -> bool {
+                Cube cubeEq(cube);
+                return cubeEq.toInteger() == num;
+            }
```
Avoid creating unnecessary copies of objects.
```c++
            [[nodiscard]] auto equals(const std::uint64_t num) const -> bool {
                return cube.toInteger() == num;
            }
```
But, even more important, the two functions here use different kinds of logic for doing the comparison. This should much rather be (which also properly works If the stored cube has more than 64bits)
```suggestion
            [[nodiscard]] auto equals(const std::uint64_t num) const -> bool {
                return cube == Cube::fromInteger(num);
            }
```

SmaranTum replied:
is there a way to get the length of occupied bits from the uint32_t ? I don't think I can overload the equals function then.

burgholzer replied:
You could just add an initial "bw" bitwidth parameter just as for the other methods taking an integer.

SmaranTum replied:
done 👍🏼

burgholzer commented on include/core/truthTable/truth_table.hpp:
```diff
@@ -112,6 +133,16 @@ namespace syrec {
+                return cubeEq.toInteger() == num;
+            }
+
+            [[nodiscard]] auto equals(const std::string& str) const -> bool {
+                Cube cubeEq(cube);
+                return Cube::fromString(str) == cubeEq;
+            }
```
avoid creating unnecessary copies.
```suggestion
            [[nodiscard]] auto equals(const std::string& str) const -> bool {
                return cube == Cube::fromString(str);
            }
```

SmaranTum replied:
done 👍🏼

burgholzer commented on include/core/truthTable/truth_table.hpp:
```diff
@@ -150,6 +181,10 @@ namespace syrec {
             return cubeMap.begin()->second.size();
         }
 
+        auto findCubeInteger(const std::uint64_t number, const std::size_t bw) -> decltype(cubeMap.cbegin()) {
```
```suggestion
        auto find(const std::uint64_t number, const std::size_t bw) -> decltype(cubeMap.cbegin()) {
```
Just a minor suggestion, since the intent already becomes clear from the parameters of the function.

SmaranTum replied:
done 👍🏼

burgholzer commented on src/algorithms/synthesis/encoding.cpp:
```diff
@@ -0,0 +1,126 @@
+        assert(tt.cubeMap.empty() || (tt.nInputs() <= std::log2(tt.cubeMap.max_size()) && tt.nInputs() <= 63U));
+
+        if (!tt.cubeMap.empty() && (tt.nInputs() > static_cast<std::size_t>(std::log2(tt.cubeMap.max_size())) || tt.nInputs() > 63U))
+            throw std::invalid_argument("Overflow!, Number of inputs more than 64");
```
This error message is wrong. 
First of all, one of the cases where this triggers is when there are more than _63_ inputs. And there is also a second case (the first condition in the if statement).
Print a proper error message here and please also cover that in a test case.

SmaranTum replied:
Apart from changing the error message to 63 inputs and removing the empty condition. what else should I do? I'm a bit confused here.

burgholzer replied:
Nothing needs to be removed at all.
There are simply to conditions that can trigger the error, i.e. `tt.nInputs() > static_cast<std::size_t>(std::log2(tt.cubeMap.max_size())` and `tt.nInputs() > 63U`.
The error message should reflect the proper value that was the cause of the exception.
Specifically whichever of `static_cast<std::size_t>(std::log2(tt.cubeMap.max_size())` and `63` is smaller.

SmaranTum replied:
done 👍🏼

burgholzer commented on src/algorithms/synthesis/encoding.cpp:
```diff
@@ -0,0 +1,126 @@
+
+        // if the truth table function is already reversible, no encoding is necessary
+        if (outputFreq.size() == tt.cubeMap.size()) {
+            return;
```
this line should be easy to cover with a test that plugs in a reversible TT and asserts that the resulting truth table is the same.

SmaranTum replied:
done 👍🏼

burgholzer commented on src/algorithms/synthesis/encoding.cpp:
```diff
@@ -0,0 +1,126 @@
+        }
+    }
+
+} // namespace syrec
```
there is a newline missing at the end of this file.

SmaranTum replied:
done 👍🏼

burgholzer commented on src/core/io/pla_parser.cpp:
```diff
@@ -29,20 +39,26 @@ namespace syrec {
 
                 if (inout.size() != 2)
-                    throw std::invalid_argument("Invalid I/O");
+                    throw std::invalid_argument("Expected exactly 2, received " + std::to_string(inout.size()));
```
Expected two .. what? there's something missing in that error message.

SmaranTum replied:
done 👍🏼

burgholzer commented on test/unittests/test_extend_tt.cpp:
```diff
@@ -10,52 +10,25 @@ class TruthTableExtend: public testing::Test {
     EXPECT_TRUE(read_pla(tt, circIdent2Bit));
 
-    EXPECT_EQ(tt.ioCube().size(), 3U);
+    EXPECT_EQ(tt.cubeMap.size(), 3U);
```
the `cubeMap` should not be accessible from outside the class as previously pointed out.
use the public interface of the TT class instead.
This also holds for all the other test files.

If it is not already, make `cubeMap`a private member of the TT class.

SmaranTum replied:
So I have a few questions on this. Initially, the cubeMap was private, however, you suggested that the extend, huffman  and augmentconstants function should not be in the truthtable class. These functions use the cubeMap member of the TT class. Since I separate them to different files, I had to make the cubeMap class public and TT class a struct. Therefore there is no public interface for the cubemap.

SmaranTum replied:
What do you suggest I do?

burgholzer replied:
TruthTable should be a class and the cubeMap should be private (as should be all functions that need not be exposed publicly).

If the other functions need some sort of access to internal structures then provide respective public access/iteration methods in the truth table class.

SmaranTum replied:
Okay, I will change this accordingly.

SmaranTum replied:
done 👍🏼

burgholzer reviewed (changes requested):
Thanks for the updates. Still a couple of things that are not ideal. I'd assume this is the last round of review necessary.

burgholzer commented on test/unittests/test_huffman.cpp:
```diff
@@ -38,40 +62,40 @@ TEST_F(TestHuff, HUFF1) {
-        EXPECT_TRUE(search != tt.cubeMap.end());
+        auto search = tt.find(in1, 2U);
+        EXPECT_TRUE(search != tt.end());
         EXPECT_TRUE(search->second.equals(std::string("0-")));
```
```suggestion
        EXPECT_TRUE(search->second.equals("0-"));
```
the cast to `std::string` is unnecessary here. please also check in all other cases/files.

SmaranTum replied:
done 👍🏼

burgholzer reviewed (changes requested):
Thanks for all the changes. This already looks much better now.
Still a couple of oversights or technicalities that need to be addressed, but very close to the finish line.

SmaranTum commented on include/core/truthTable/truth_table.hpp:
```diff
@@ -42,6 +41,20 @@ namespace syrec {
                 std::copy(first, last, std::back_inserter(cube));
             }
 
+            static auto getValue(const char& c) -> std::optional<bool> {
```
done 👍🏼

SmaranTum commented on include/core/truthTable/truth_table.hpp:
```diff
@@ -133,14 +144,12 @@ namespace syrec {
-            [[nodiscard]] auto equals(const std::uint64_t num) const -> bool {
-                Cube cubeEq(cube);
-                return cubeEq.toInteger() == num;
+            [[nodiscard]] auto equals(const std::uint64_t num, const std::uint64_t bw) const -> bool {
```
done 👍🏼

SmaranTum commented on include/core/truthTable/truth_table.hpp:
```diff
@@ -133,14 +144,12 @@ namespace syrec {
             [[nodiscard]] auto equals(const std::string& str) const -> bool {
-                Cube cubeEq(cube);
-                return Cube::fromString(str) == cubeEq;
+                return cube == (Cube::fromString(str)).cube;
```
done 👍🏼

SmaranTum commented on include/core/truthTable/truth_table.hpp:
```diff
@@ -133,14 +144,12 @@ namespace syrec {
-                Cube cubeEq(cube);
-                return cubeEq.toInteger() == num;
+            [[nodiscard]] auto equals(const std::uint64_t num, const std::uint64_t bw) const -> bool {
+                return cube == (Cube::fromInteger(num, bw)).cube;
```
done 👍🏼

SmaranTum commented on include/core/truthTable/truth_table.hpp:
```diff
@@ -194,6 +252,10 @@ namespace syrec {
         }
 
+        auto insertNode(CubeMap::node_type nh) -> void {
+            cubeMap.insert(std::move(nh));
```
Can it be used? emplace function does not have any thing related to node_type

burgholzer replied:
I mean the easiest way to find that out is to just try it. `emplace` merely forwards the given arguments, so I would expect it to work straight out of the box.

SmaranTum replied:
I tried it out. Its telling no matching function for emplace

burgholzer replied:
fair enough. then this is fine as it is 👍🏼

SmaranTum replied:
👍🏼

SmaranTum commented on include/core/truthTable/truth_table.hpp:
```diff
@@ -194,6 +252,10 @@ namespace syrec {
             cubeMap.try_emplace(std::move(input), std::move(output));
         }
 
+        auto insertNode(CubeMap::node_type nh) -> void {
```
done 👍🏼

SmaranTum commented on src/algorithms/synthesis/encoding.cpp:
```diff
@@ -4,54 +4,52 @@ namespace syrec {
 
-        if (!tt.cubeMap.empty() && (tt.nInputs() > static_cast<std::size_t>(std::log2(tt.cubeMap.max_size())) || tt.nInputs() > 63U))
-            throw std::invalid_argument("Overflow!, Number of inputs more than 64");
+        TruthTable newCubeMap{};
```
done 👍🏼

SmaranTum commented on src/algorithms/synthesis/encoding.cpp:
```diff
@@ -98,13 +96,13 @@ namespace syrec {
-        for (auto& [input, output]: tt.cubeMap) {
-            output = encoding[output];
+        for (auto& [input, output]: tt) {
+            tt[input] = encoding[output];
```
I tried this out yesterday, its telling "No viable overloaded '='"

burgholzer replied:
The answer for that problem is pretty simple. Only one of the two necessary `begin` and `end` functions are defined in the `TruthTable` class. The ones defined right now return a `const_iterator`. You also need to add non-const versions, i.e.
```c++
        [[nodiscard]] auto begin() -> decltype(cubeMap.begin()) {
            return cubeMap.begin();
        }
```
and
```c++
        [[nodiscard]] auto end() -> decltype(cubeMap.end()) {
            return cubeMap.end();
        }
```

SmaranTum replied:
👍🏼

SmaranTum commented on src/algorithms/synthesis/encoding.cpp:
```diff
@@ -4,54 +4,52 @@ namespace syrec {
             // move all the complete cubes to the new cube map
-            for (auto& completeInput: completeInputs) {
-                newCubeMap.try_emplace(std::move(completeInput), output);
+            for (auto const& completeInput: completeInputs) {
+                newCubeMap.insert(completeInput, output);
```
done 👍🏼

burgholzer reviewed (approved):
LGTM 👍🏼  Thanks for all the effort put in. I believe it was definitely worth it.

I'll update the title and the description of this PR and merge it after all CI checks pass! 🥳

SmaranTum commented:
> LGTM 👍🏼 Thanks for all the effort put in. I believe it was definitely worth it.
> 
> I'll update the title and the description of this PR and merge it after all CI checks pass! partying_face

Yes, it was worth it. Tbh, I learnt a lot from this. Thank you.
